src/AzureAuth/Ado/AdoToken.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AzureAuth.Ado
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Microsoft.Office.Lasso.Interfaces;
+
+    /// <summary>
+    /// Business logic for either getting an ADO PAT from <see cref="IEnv"/>
+    /// or doing normal AAD Authentication.
+    /// </summary>
+    public static class AdoToken
+    {
+        private static readonly IEnumerable<string> PatEnvVars = new[]
+        {
+            EnvVars.AdoPat,
+            EnvVars.SystemAccessToken,
+        };
+
+        /// <summary>
+        /// Get either an ADO PAT from the environment. If one is not set, then do normal ADO AAD Auth.
+        /// </summary>
+        /// <param name="env">The <see cref="IEnv"/> to use for interacting with the environment.</param>
+        /// <returns>A PAT from the env if set, otherwise null.</returns>
+        public static PatResult PatFromEnv(IEnv env)
+        {
+            foreach (var envVar in PatEnvVars)
+            {
+                var pat = env.Get(envVar);
+                var exists = !string.IsNullOrEmpty(pat);
+
+                if (exists)
+                {
+                    return new ()
+                    {
+                        Exists = exists,
+                        Value = pat,
+                        EnvVarSource = envVar,
+                    };
+                }
+            }
+
+            return new ()
+            {
+                Exists = false,
+            };
+        }
+
+        /// <summary>
+        /// The result struct for retrieving a PAT out of the environment.
+        /// </summary>
+        public record PatResult
+        {
+            /// <summary>
+            /// Gets a value indicating whether a PAT was found in the env.
+            /// </summary>
+            public bool Exists { get; init; }
+
+            /// <summary>
+            /// Gets the source env var that contained the PAT.
+            /// </summary>
+            public string EnvVarSource { get; init; }
+
+            /// <summary>
+            /// Gets the PAT value itself.
+            /// </summary>
+            public string Value { get; init; }
+        }

```

Comment:
Consider moving this into a PatResult.cs file.

Reply:
Not having a strong C# background I'm genuinely curious about this. Why? (Or, conversely, why not?)

From my understanding of the structure here the `PatResult` is the main value that's meant to be passed around. The `AdoToken` class is effectively acting like a static method which constructs this value. I see `PatResult` and `AdoToken` as being intimately linked, but maybe they don't need to be?

Is there a structural reason to do this or is it just following the principal of "one class per file"?

Reply:
The fact they are intimately linked, is why I chose not to put this into another class. The Pat from env result is an implementation detail. I'd like to make it private. But we run into visibility issues in the Tests, since I can't just put the test for this directly into this file... like we could in Rust.

Alternatively to a Record - this could return a named tuple. But this is a little uglier IMO, though it would work and then no private type would be needed.

Reply:
@shalinikhare27 is this reasoning sound enough for you to leave this inside the AdoToken class?